core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -3865,6 +3865,18 @@ object ZIO extends ZIOCompanionPlatformSpecific {
   def services[A: Tag, B: Tag, C: Tag, D: Tag]: URIO[Has[A] with Has[B] with Has[C] with Has[D], (A, B, C, D)] =
     ZIO.access(r => (r.get[A], r.get[B], r.get[C], r.get[D]))
 
+  /**
+   * Effectfully accesses the specified service in the environment of the effect.
+   *
+   * Especially, useful for creating "accessor" methods on Services' companion objects.
+   *
+   * {{{
+   * def foo(int: Int) = ZIO.serviceWith[Foo](_.foo(int))
+   * }}}
+   */
+  def serviceWith[Service]: ServiceWithPartiallyApplied[Service] =
```

Comment:
Apologies for the naive question, but why is the "partially applied" pattern used in this scenario? I was under the impression that it helps with type inference if the method is supposed to walk, talk and quack as though it had multiple parameter lists.

Reply:
Unfortunately, one cannot _partially apply_ type parameters. We need to use this pattern if we want to be able to call this method as such:
```scala
ZIO.serviceWith[Foo](_.foo(int))
```
If it didn't use this awkward PartiallyApplied class trick, we'd have to specify the error and return types:

```scala
ZIO.serviceWith[Foo, Nothing, Int](_.foo(int))
```

Reply:
Gotcha. In an alternative universe, would an alternative Scala allow something like this?
```
def serviceWith[Service][E, A](f: Service => ZIO[Has[Service], E, A])
```
If so, has such syntax been proposed with the keepers of Scala?

Reply:
> Gotcha. In an alternative universe, would an alternative Scala allow something like this?
> 
> ```
> def serviceWith[Service][E, A](f: Service => ZIO[Has[Service], E, A])
> ```
> 
> If so, has such syntax been proposed with the keepers of Scala?

In scala3 you have [named type arguments](https://dotty.epfl.ch/docs/reference/other-new-features/named-typeargs)

If I'm not mistaken, it should allow you to do something like:
```scala
ZIO.serviceWith[Service = Foo](_.foo(int))
```